p2p/src/types.rs

```diff
@@ -106,51 +112,72 @@ impl<T> From<mpsc::TrySendError<T>> for Error {
 	}
 }
 
-#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
-pub struct PeerAddr(pub SocketAddr);
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum PeerAddr {
+	Socket(SocketAddr),
+	I2p(I2pSocketAddr),
+}
 
 impl Writeable for PeerAddr {
 	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
-		match self.0 {
-			SocketAddr::V4(sav4) => {
+		match self {
+			PeerAddr::Socket(SocketAddr::V4(sav4)) => {
 				ser_multiwrite!(
 					writer,
 					[write_u8, 0],
 					[write_fixed_bytes, &sav4.ip().octets().to_vec()],
 					[write_u16, sav4.port()]
 				);
 			}
-			SocketAddr::V6(sav6) => {
+			PeerAddr::Socket(SocketAddr::V6(sav6)) => {
 				writer.write_u8(1)?;
 				for seg in &sav6.ip().segments() {
 					writer.write_u16(*seg)?;
 				}
 				writer.write_u16(sav6.port())?;
 			}
+			PeerAddr::I2p(i2p_addr) => {
+				ser_multiwrite!(
+					writer,
+					[write_u8, 2],
+					[write_bytes, &i2p_addr.dest().to_string().into_bytes()],
+					[write_u16, i2p_addr.port()]
+				);
+			}
 		}
 		Ok(())
 	}
 }
 
 impl Readable for PeerAddr {
 	fn read(reader: &mut dyn Reader) -> Result<PeerAddr, ser::Error> {
-		let v4_or_v6 = reader.read_u8()?;
-		if v4_or_v6 == 0 {
+		let addr_format = reader.read_u8()?;
+		if addr_format == 0 {
 			let ip = reader.read_fixed_bytes(4)?;
 			let port = reader.read_u16()?;
-			Ok(PeerAddr(SocketAddr::V4(SocketAddrV4::new(
+			Ok(PeerAddr::Socket(SocketAddr::V4(SocketAddrV4::new(
 				Ipv4Addr::new(ip[0], ip[1], ip[2], ip[3]),
 				port,
 			))))
-		} else {
+		} else if addr_format == 1 {
 			let ip = try_iter_map_vec!(0..8, |_| reader.read_u16());
 			let port = reader.read_u16()?;
-			Ok(PeerAddr(SocketAddr::V6(SocketAddrV6::new(
+			Ok(PeerAddr::Socket(SocketAddr::V6(SocketAddrV6::new(
 				Ipv6Addr::new(ip[0], ip[1], ip[2], ip[3], ip[4], ip[5], ip[6], ip[7]),
 				port,
 				0,
 				0,
 			))))
+		} else if addr_format == 2 {
+			let addr = reader.read_bytes_len_prefix()?;
```

Comment:
`read_bytes_len_prefix` looks dangerous, I'd change it to accept `max_len` as a parameter to be validated before reading the content, can send a tiny pr.

Reply:
Thought the same at first but it actually refuses to read more than 100K. And then that would get you banned right away at the first address read.